### ISSUE TITLE
Github test account disabled

### ISSUE THREAD

**Issue by cseed** [hail-is/hail#4517]
- [ ] There should be a test in the tests that verify that the account is operational, so if it disabled, we get an informative error message.

 - [ ] There should be documentation/a playbook about how to get Github unstuck when this happens.

 - [ ] We should mock Github so we don't rely on it during the tests, which makes me sad because yay integration tests, and what do we do when the Github API changes/breaks/doesn't behave the way we expect?

Assigning to @danking since he's been through the first two, but maybe someone else can handle the third.

**Comment by cseed** [comment]
This is in response to this failure: https://storage.googleapis.com/hail-ci-0-1/ci/7aa524504b8bafe0a4af859e73bc4f9efdaa052c/39a94649482a2512a7a514e6084c5b84f48b8205/index.html

Which might be mis-diagnosed.

**Comment by danking** [comment]
Strange, the bad point seems to be:
```
++ mktemp -d
+ REPO_DIR=/tmp/tmp.2v53NEHcHZ
+ cp test-repo/hail-ci-build-image test-repo/hail-ci-build.sh test-repo/hail-ci-deploy.sh /tmp/tmp.2v53NEHcHZ
/tmp/tmp.2v53NEHcHZ /hail/repo/ci
+ pushd /tmp/tmp.2v53NEHcHZ
+ git init
Initialized empty Git repository in /tmp/tmp.2v53NEHcHZ/.git/
+ git config user.email ci-automated-tests@broadinstitute.org
+ git config user.name ci-automated-tests
+ set +x
+ git add hail-ci-build-image hail-ci-build.sh hail-ci-deploy.sh
+ git commit -m 'inital commit'
[master (root-commit) da0ddab] inital commit
 3 files changed, 26 insertions(+)
 create mode 100644 hail-ci-build-image
 create mode 100644 hail-ci-build.sh
 create mode 100644 hail-ci-deploy.sh
+ git push origin master:master
error: RPC failed; HTTP 404 curl 22 The requested URL returned error: 404 Not Found
fatal: The remote end hung up unexpectedly
fatal: The remote end hung up unexpectedly
Everything up-to-date
+ cleanup
```
the `git puts origin master:master`. Hidden from the logs is the URL because it contains a token.

**Comment by cseed** [comment]
It looks like the repo create succeeded.  Is there a race condition and it isn't ready yet?

**Comment by danking** [comment]
So the sequence of events is:
 - create repo on GH (via REST call)
 - initialize local repository (no GH comms)
 - add remote (no GH comms)
 - push to GH

(see [test-locally.sh](https://github.com/hail-is/hail/blob/master/ci/test-locally.sh#L31-L53), which is called by test-cluster.sh, after exposing its pod via nginx)

We're failing on that last step.

It appears that the repository creation succeeded, the response to that REST call was a successful looking JSON object (which indicates a create repository with the expected name, `REPO_NAME=ci-test-p4a9fxo7`) [1].

It is unclear if repository deletion succeeded [2], I should find a way to make curl print the HTTP status code returned by cluster deletion. FWIW, the repository appears to be deleted now (https://github.com/hail-ci-test/ci-test-p4a9fxo7), and, AFAIK, no other part of our system deletes these repositories.


[1]:
```
{
  "id": 152339517,
  "node_id": "MDEwOlJlcG9zaXRvcnkxNTIzMzk1MTc=",
  "name": "ci-test-p4a9fxo7",
  "full_name": "hail-ci-test/ci-test-p4a9fxo7",
  "private": false,
  "owner": {
    "login": "hail-ci-test",
    "id": 43152710,
    "node_id": "MDEyOk9yZ2FuaXphdGlvbjQzMTUyNzEw",
    "avatar_url": "https://avatars1.githubusercontent.com/u/43152710?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/hail-ci-test",
    "html_url": "https://github.com/hail-ci-test",
    "followers_url": "https://api.github.com/users/hail-ci-test/followers",
    "following_url": "https://api.github.com/users/hail-ci-test/following{/other_user}",
    "gists_url": "https://api.github.com/users/hail-ci-test/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/hail-ci-test/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/hail-ci-test/subscriptions",
    "organizations_url": "https://api.github.com/users/hail-ci-test/orgs",
    "repos_url": "https://api.github.com/users/hail-ci-test/repos",
    "events_url": "https://api.github.com/users/hail-ci-test/events{/privacy}",
    "received_events_url": "https://api.github.com/users/hail-ci-test/received_events",
    "type": "Organization",
    "site_admin": false
  },
  "html_url": "https://github.com/hail-ci-test/ci-test-p4a9fxo7",
  "description": null,
  "fork": false,
  "url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7",
  "forks_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/forks",
  "keys_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/keys{/key_id}",
  "collaborators_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/collaborators{/collaborator}",
  "teams_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/teams",
  "hooks_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/hooks",
  "issue_events_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/issues/events{/number}",
  "events_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/events",
  "assignees_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/assignees{/user}",
  "branches_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/branches{/branch}",
  "tags_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/tags",
  "blobs_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/git/blobs{/sha}",
  "git_tags_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/git/tags{/sha}",
  "git_refs_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/git/refs{/sha}",
  "trees_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/git/trees{/sha}",
  "statuses_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/statuses/{sha}",
  "languages_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/languages",
  "stargazers_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/stargazers",
  "contributors_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/contributors",
  "subscribers_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/subscribers",
  "subscription_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/subscription",
  "commits_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/commits{/sha}",
  "git_commits_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/git/commits{/sha}",
  "comments_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/comments{/number}",
  "issue_comment_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/issues/comments{/number}",
  "contents_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/contents/{+path}",
  "compare_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/compare/{base}...{head}",
  "merges_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/merges",

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  7030  100  6999  100    31   9306     41 --:--:-- --:--:-- --:--:--  9319
  "archive_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/{archive_format}{/ref}",
  "downloads_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/downloads",
  "issues_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/issues{/number}",
  "pulls_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/pulls{/number}",
  "milestones_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/milestones{/number}",
  "notifications_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/notifications{?since,all,participating}",
  "labels_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/labels{/name}",
  "releases_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/releases{/id}",
  "deployments_url": "https://api.github.com/repos/hail-ci-test/ci-test-p4a9fxo7/deployments",
  "created_at": "2018-10-10T00:32:59Z",
  "updated_at": "2018-10-10T00:32:59Z",
  "pushed_at": "2018-10-10T00:33:00Z",
  "git_url": "git://github.com/hail-ci-test/ci-test-p4a9fxo7.git",
  "ssh_url": "git@github.com:hail-ci-test/ci-test-p4a9fxo7.git",
  "clone_url": "https://github.com/hail-ci-test/ci-test-p4a9fxo7.git",
  "svn_url": "https://github.com/hail-ci-test/ci-test-p4a9fxo7",
  "homepage": null,
  "size": 0,
  "stargazers_count": 0,
  "watchers_count": 0,
  "language": null,
  "has_issues": true,
  "has_projects": true,
  "has_downloads": true,
  "has_wiki": true,
  "has_pages": false,
  "forks_count": 0,
  "mirror_url": null,
  "archived": false,
  "open_issues_count": 0,
  "license": null,
  "forks": 0,
  "open_issues": 0,
  "watchers": 0,
  "default_branch": "master",
  "permissions": {
    "admin": true,
    "push": true,
    "pull": true
  },
  "allow_squash_merge": true,
  "allow_merge_commit": true,
  "allow_rebase_merge": true,
  "organization": {
    "login": "hail-ci-test",
    "id": 43152710,
    "node_id": "MDEyOk9yZ2FuaXphdGlvbjQzMTUyNzEw",
    "avatar_url": "https://avatars1.githubusercontent.com/u/43152710?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/hail-ci-test",
    "html_url": "https://github.com/hail-ci-test",
    "followers_url": "https://api.github.com/users/hail-ci-test/followers",
    "following_url": "https://api.github.com/users/hail-ci-test/following{/other_user}",
    "gists_url": "https://api.github.com/users/hail-ci-test/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/hail-ci-test/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/hail-ci-test/subscriptions",
    "organizations_url": "https://api.github.com/users/hail-ci-test/orgs",
    "repos_url": "https://api.github.com/users/hail-ci-test/repos",
    "events_url": "https://api.github.com/users/hail-ci-test/events{/privacy}",
    "received_events_url": "https://api.github.com/users/hail-ci-test/received_events",
    "type": "Organization",
    "site_admin": false
  },
  "network_count": 0,
  "subscribers_count": 0
}
```

[2]:
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
```

**Comment by danking** [comment]
It is possible there is a race condition, though I have not witnessed this before.

In fact, it seems rather reasonable that GitHub had some intermittent slow down that delayed repository creation or ability to find said repository temporarily.

**Comment by danking** [comment]
Also, I think you've referenced the wrong log, the one from your original reporting email is: https://storage.googleapis.com/hail-ci-0-1/ci/7aa524504b8bafe0a4af859e73bc4f9efdaa052c/414f3f183bd5f2ec04e1c732522cbc0b8b1fca31/job-log

The log you referenced appears to be failing because another process (presumably `batch`) has already bound to port 5000. Perhaps `batch`'s `hail-ci-build.sh` is not properly cleaning up processes when it exits.

**Comment by danking** [comment]
The repository in question was created at `2018-10-10T00:32:59Z`. GitHub indicates [no system failures](https://status.github.com/messages) on October the tenth or the ninth.

**Comment by danking** [comment]
So there appear to be to distinct issues:

 - failure to push the initial data to a repository due to a 404 https://storage.googleapis.com/hail-ci-0-1/ci/7aa524504b8bafe0a4af859e73bc4f9efdaa052c/414f3f183bd5f2ec04e1c732522cbc0b8b1fca31/job-log

 - 404s due to someone else already bound to port 5000 (maybe `batch`?) https://storage.googleapis.com/hail-ci-0-1/ci/7aa524504b8bafe0a4af859e73bc4f9efdaa052c/39a94649482a2512a7a514e6084c5b84f48b8205/index.html
```
Traceback (most recent call last):
  File "ci/ci.py", line 372, in <module>
    app.run(host='0.0.0.0', threaded=False)
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/site-packages/flask/app.py", line 943, in run
    run_simple(host, port, self, **options)
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/site-packages/werkzeug/serving.py", line 814, in run_simple
    inner()
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/site-packages/werkzeug/serving.py", line 774, in inner
    fd=fd)
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/site-packages/werkzeug/serving.py", line 666, in make_server
    passthrough_errors, ssl_context, fd=fd)
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/site-packages/werkzeug/serving.py", line 577, in __init__
    self.address_family), handler)
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/socketserver.py", line 449, in __init__
    self.server_bind()
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/home/hail/.conda/envs/hail-ci/lib/python3.7/socketserver.py", line 463, in server_bind
    self.socket.bind(self.server_address)
OSError: [Errno 98] Address already in use
```

**Comment by danking** [comment]
For the latter issue I'm creating #4531.

**Comment by danking** [comment]
I'm fairly certain this was an intermittent race condition in GitHub that we have no control over. Closing. Re-open if we see it again.

**Comment by danking** [comment]
Created https://github.com/hail-is/hail/issues/4533 to track testing for account liveliness

There's no good solution to bullet two, I just email GitHub support and ask them to unblock. I guess this page is now the documentation for that.

**Comment by danking** [comment]
Bullet three is probably worth doing but low priority for me.
